persistence/sql/persister_grant_jwk.go

```diff
@@ -0,0 +1,196 @@
+package sql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gobuffalo/pop/v5"
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/ory/hydra/oauth2/trust"
+	"github.com/ory/x/errorsx"
+
+	"github.com/ory/x/sqlcon"
+)
+
+var _ trust.GrantManager = &Persister{}
+
+const scopeSeparator = " "
+
+func (p *Persister) CreateGrant(ctx context.Context, g trust.Grant, publicKey jose.JSONWebKey) error {
+	// add key, if it doesn't exist
+	if _, err := p.GetKey(ctx, g.PublicKey.Set, g.PublicKey.KeyID); err != nil {
+		if errorsx.Cause(err) != sqlcon.ErrNoRows {
+			return err
+		}
+
+		if err = p.AddKey(ctx, g.PublicKey.Set, &publicKey); err != nil {
+			return err
+		}
+	}
+
+	data := p.sqlDataFromJWTGrant(g)
+
+	return sqlcon.HandleError(p.Connection(ctx).Create(&data))
+}
+
+func (p *Persister) GetConcreteGrant(ctx context.Context, id string) (trust.Grant, error) {
+	var data trust.SQLData
+	if err := p.Connection(ctx).Where("id = ?", id).First(&data); err != nil {
+		return trust.Grant{}, sqlcon.HandleError(err)
+	}
+
+	return p.jwtGrantFromSQlData(data), nil
+}
+
+func (p *Persister) DeleteGrant(ctx context.Context, id string) error {
+	grant, err := p.GetConcreteGrant(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	return p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
+		if err := p.Connection(ctx).Destroy(&trust.SQLData{ID: grant.ID}); err != nil {
+			return sqlcon.HandleError(err)
+		}
+
+		return p.DeleteKey(ctx, grant.PublicKey.Set, grant.PublicKey.KeyID)
+	})
+}
+
+func (p *Persister) GetGrants(ctx context.Context, limit, offset int, optionalIssuer string) ([]trust.Grant, error) {
+	grantsData := make([]trust.SQLData, 0)
+
+	query := p.Connection(ctx).Paginate(offset/limit+1, limit).Order("id")
+	if optionalIssuer != "" {
+		query = query.Where("issuer = ?", optionalIssuer)
+	}
+
+	if err := query.All(&grantsData); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	grants := make([]trust.Grant, 0, len(grantsData))
+	for _, data := range grantsData {
+		grants = append(grants, p.jwtGrantFromSQlData(data))
+	}
+
+	return grants, nil
+}
+
+func (p *Persister) CountGrants(ctx context.Context) (int, error) {
+	n, err := p.Connection(ctx).Count(&trust.SQLData{})
+	return n, sqlcon.HandleError(err)
+}
+
+func (p *Persister) GetPublicKey(ctx context.Context, issuer string, subject string, keyId string) (*jose.JSONWebKey, error) {
+	var data trust.SQLData
+	query := p.Connection(ctx).
+		Where("issuer = ?", issuer).
+		Where("subject = ?", subject).
+		Where("key_id = ?", keyId)
+	if err := query.First(&data); err != nil {
+		return nil, sqlcon.HandleError(err)
+	}
+
+	keySet, err := p.GetKey(ctx, data.KeySet, keyId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &keySet.Keys[0], nil
```

Comment:
Is it ensured that we always have one key?

Reply:
This uses the `GetKey` function, which takes care of returning an error if it does not find the key, or if it fails at unmarshaling. So I would say this is pretty safe as it will always return at least one non-nil key: `Keys: []jose.JSONWebKey{c}`.